### DISCUSSION TITLE
Streaming JSON results in real time with HttpClient.GetFromJsonAsAsyncEnumerable doesn't work in the WASM stack

### DISCUSSION THREAD

**Discussion by NetherGranite** [dotnet/runtime#110287] [Q&A] [answered]
Currently, `HttpClient.GetFromJsonAsAsyncEnumerable` doesn't return any results until a significant amount of the JSON response has been buffered. Instead, it waits until the full response has been received. I believe this is behavior of `JsonSerializer.DeserializeAsyncEnumerable`, but I haven't confirmed for myself. After a good hour of research and looking through random Stack Overflow and GitHub, the two workarounds I was able to find are to 1) lower the default buffer size of the JSON deserializer to a miniscule value, which didn't work for me, and 2) flush the HTTP response body every time the async enumerable returns an in-progress task, something I'm not sure how to do properly and that's very tedious regardless.

Edit: I should clarify that this is a client-side issue. Assuming the server is properly streaming results and the client browser is receiving them as early as possible, `HttpClient.GetFromJsonAsAsyncEnumerable` still waits until all results have been fetched to complete.

My questions are these:

- Is there a recommended workaround or even proper solution to this that's just not well-known?
- What part of the JSON deserializer causes this behavior?
- Is there a reason why this is the default behavior out of the box?

**Comment by MihaZupan** [comment]
`GetFromJsonAsAsyncEnumerable` shouldn't be doing any such buffering that would delay when you're able to see results.

> 2) flush the HTTP response body every time the async enumerable returns an in-progress task

That's up to the server producing the data. If the server doesn't proactively flush then of course the client can't see the result right away.

Are you able to create a minimal runnable repro that reproduces the issue?

**Reply by NetherGranite** [reply]
> That's up to the server producing the data. If the server doesn't proactively flush then of course the client can't see the result right away.

Sorry, I should have clarified that this is a client-side issue and assumes the server is properly returning results as soon as they're available.

> Are you able to create a minimal runnable repro that reproduces the issue?

Yes, here is a repo that adds a very small amount of sample code to the standard Blazor Web App template demonstrating the issue: [AsyncEnumerableIssueRepro](https://github.com/NetherGranite/AsyncEnumerableIssueRepro). Just run the app and click on the "Load Items" button on the home page and see that while the server returns results in real time in the Network tab of dev tools, the `IAsyncEnumerable` from `HttpClient.GetFromJsonAsAsyncEnumerable` returns all the results at once after all of them are fetched.

**Reply by MihaZupan** [reply, marked as answer]
Ah, the client in this case is running via WebAssembly on the client browser itself, that uses a different networking stack.
There, responses are not streamed by default, and you have to opt back in via `SetBrowserResponseStreamingEnabled`.

Changing your sample from
```c#
var itemStream = HttpClient.GetFromJsonAsAsyncEnumerable<string>("/api/items");
```
to
```c#
var request = new HttpRequestMessage(HttpMethod.Get, "/api/items");
request.SetBrowserResponseStreamingEnabled(true);
using var response = await HttpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead);
var itemStream = response.Content.ReadFromJsonAsAsyncEnumerable<string>();
```
gets you the streaming behavior as expected.

There's some more discussion around this in #77904

**Reply by NetherGranite** [reply]
Oh my goodness... This was impossible to track down with Google searches. I will try to update this post with some keywords to help others in the future. Thank you!

**Comment by WeihanLi** [comment]
https://github.com/dotnet/runtime/issues/102113

I met this issue before when using the response compression, not sure if the same case, hope it helps

**Reply by NetherGranite** [reply]
Is compression enabled by default by chance? I haven't done anything to enable it

Edit: This was resolved, thank you!
